File: 46.全排列[重复训练].cpp
// https://leetcode.cn/problems/permutations/
// 题解：DFS，理解当前for循环步骤执行完，才能执行下一for循环，故数组[1,x,x]先遍历完全。
class Solution {
public:
    ;
    vector<vector<int>> permute(vector<int> &nums) {
        vector<vector<int>> res;
        dfs(nums, res, 0);
        return res;
    }

    void dfs(vector<int> nums, vector<vector<int>> &res, int left) {
        int n = nums.size();

        if(left == n-1) {
            res.emplace_back(nums);
        } else {
            for(int i = left; i < n; i++) {
                swap(nums[i], nums[left]);
                dfs(nums, res, left+1);
            }
        }
    }
};

/*
[1,2,3]

[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
*/

File: 46.全排列[重复训练].java
// https://leetcode.cn/problems/permutations/
// 题解：深度优先遍历DFS，注意vis[]标记数组的使用，以及ArrayList的深浅拷贝。
class Solution {
    private List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        int n = nums.length;
        boolean[] vis = new boolean[n];
        List<Integer> perm_arr = new ArrayList<>();

        dfs(nums, perm_arr, vis, 0);
        return ans;
    }

    void dfs(int[] nums, List<Integer> perm_arr, boolean[] vis, int i) {
        if(i == nums.length) {
            ans.add(new ArrayList<>(perm_arr)); // deep copy.
            return;
        }

        for(int k = 0; k < nums.length; k++) {
            if(!vis[k]) {
                vis[k] = true;
                perm_arr.add(nums[k]);
                dfs(nums, perm_arr, vis, i + 1);
                perm_arr.remove(perm_arr.size() - 1);
                vis[k] = false;
            }
        }
    }
}

/*
[1,2,3]

[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
*/